### PR TITLE
Fix module name handling

### PR DESCRIPTION
**Pull request by pablocarle**
## Proposed changes

String in ASCII / UTF-8 that reached the quick-js code in js_load_file function was corrupted.
It's not yet clear why or how to trigger it (I think it's possible most installations go through the upper if branch here (https://github.com/zowe/zowe-common-c/blob/a924bdd15b32a6a7e146619fa927dca5633ddd2e/c/embeddedjs.c#L1990) but when it goes in the lower branch, the string handling can corrupt the UTF-8, making a call to js_load_file with a shorter string and failing in finding the file.

The PR refactors the implementation to start with a fresh memory and copy the UTF-8/ASCII path in all cases.

This PR addresses Issue: https://github.com/zowe/zowe-install-packaging/issues/3178

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)

## PR Checklist
Please delete options that are not relevant.
- [ ] If the changes in this PR are meant for the next release / mainline, this PR targets the "staging" branch.
- [ ] My code follows the style guidelines of this project (see: [Contributing guideline](https://github.com/zowe/zss/blob/v2.x/master/CONTRIBUTING.md))
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] New and existing unit tests pass locally with my changes
- [ ] video or image is included if visual changes are made
- [ ] Relevant update to CHANGELOG.md
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works, or describe a test method below

### CODE REVIEW

File: c/embeddedjs.c
Comment (1000TurquoisePogs) [inline c/embeddedjs.c]: could this be a `char nameToLoad[modNameLen+4]`? I just don't see a free, so this will add up.
Also, `safeMalloc()` rather than `malloc()` is preferred.

Comment (pablocarle) [reply c/embeddedjs.c]: My mistake about the free, I'll add the safeMalloc use. Thanks

File: c/embeddedjs.c
Comment (1000TurquoisePogs) [inline c/embeddedjs.c]: safeFree() goes with safeMalloc(), sorry for not explaining.

https://github.com/zowe/zowe-common-c/blob/v2.x/staging/h/alloc.h#L51-L77

File: c/embeddedjs.c
Comment (JoeNemo) [reply c/embeddedjs.c]: The char[modeNameLen+4] is a stack allocation, and alloca().

https://linux.die.net/man/3/alloca   

So it's not an issue about freeing unless the function returns before the load completes.  But I doubt that because these are synchronous API's.

**Comment by JoeNemo** [comment]
In general, I am not clear on what is being fixed here.   Was the allocation length wrong, and more data got copied than should have been?  And is it still wrong but moved to a heap allocation with safeMalloc/safeFree so it didn't trash the stack.   

I would feel better if there were a case to reproduce this.

File: c/embeddedjs.c
Comment (pablocarle) [reply c/embeddedjs.c]: > In general, I am not clear on what is being fixed here. Was the allocation length wrong, and more data got copied than should have been? And is it still wrong but moved to a heap allocation with safeMalloc/safeFree so it didn't trash the stack.
> 
> I would feel better if there were a case to reproduce this.

I'm no C expert, but I think the issue is the declaration of the array within the `else` block, you can see in this screenshot the result of some logging I added:

```c
      char asciiExtendedName[modNameLen+3+1];
      memcpy(asciiExtendedName,moduleName,modNameLen);
      fprintf(f, "asciiExtendedName1: '%s'", asciiExtendedName);
      memcpy(asciiExtendedName+modNameLen,asciiDotJS,4);
      fprintf(f, "asciiExtendedName2: '%s'", asciiExtendedName);
      nameToLoad = asciiExtendedName;
    }
    fprintf(f, "nameToLoad: '%s'", nameToLoad);
    buf = js_load_file(ctx, &buf_len, nameToLoad);
```

image
[embeddedjs.log](https://github.com/zowe/zowe-common-c/files/11675249/embeddedjs.log)

This issue is fixed in our installation even with moving the array declaration outside the if block (`      char asciiExtendedName[modNameLen+3+1];`) if this is preferable I can definitely do that.

Comment (pablocarle) [reply c/embeddedjs.c]: `7D` is the apostrophe in ebcdic, so the last use of the `nameToLoad` is wrong in the hex

**Comment by pablocarle** [comment]
> In general, I am not clear on what is being fixed here. Was the allocation length wrong, and more data got copied than should have been? And is it still wrong but moved to a heap allocation with safeMalloc/safeFree so it didn't trash the stack.
> 
> I would feel better if there were a case to reproduce this.

I replied but for some reason it got in a conversation, it's here: https://github.com/zowe/zowe-common-c/pull/380#discussion_r1221273743

File: c/embeddedjs.c
Comment (JoeNemo) [reply c/embeddedjs.c]: Can you modify the commit to fix things this way.   Safemalloc/safe free was not the issue.   I will be back home in about 8-9 hours and can look then

Comment (pablocarle) [reply c/embeddedjs.c]: yes, done, thanks

**Review by 1000TurquoisePogs — APPROVED** [review]
This seems good to me, I'd like to wait for Joe to finalize his review but I'm approving to not hold back any progress on RC2.

File: c/embeddedjs.c
Comment (ifakhrutdinov) [inline c/embeddedjs.c]: Could you use `sizeof(asciiDotJS)` instead of `+ 3`. I know we'll waste 1 byte, but this would be clearer.

File: c/embeddedjs.c
Comment (ifakhrutdinov) [inline c/embeddedjs.c]: `memcpy(nameToLoad + modNameLen, asciiDotJS, sizeof(asciiDotJS));`

File: c/embeddedjs.c
Comment (ifakhrutdinov) [inline c/embeddedjs.c]: I'd initialize this to 0, you can do
```c
char nameToLoad[modNameLen + sizeof(asciiDotJS) + 1] = {0};
```

**Review by ifakhrutdinov — COMMENTED** [review]
See my comments. The changes look correct but I wouldn't use magic constants.

File: c/embeddedjs.c
Comment (pablocarle) [reply c/embeddedjs.c]: done, thanks

File: c/embeddedjs.c
Comment (pablocarle) [reply c/embeddedjs.c]: done, thanks

File: c/embeddedjs.c
Comment (pablocarle) [reply c/embeddedjs.c]: done, thanks

**Comment by struga0258** [comment]
Build is failing with the following error: variable-sized object may not be initialized
    char nameToLoad[modNameLen + sizeof(asciiDotJS) + 1] = {0};

**Comment by pablocarle** [comment]
> Build is failing with the following error: variable-sized object may not be initialized char nameToLoad[modNameLen + sizeof(asciiDotJS) + 1] = {0};

Yes, thank you I already pushed the fix but I'm trying to figure out why I don't see it

**Comment by ifakhrutdinov** [comment]
Looks fine. Could you please squash the commits?

**Comment by ifakhrutdinov** [comment]
All, if the commit history doesn't get cleaned up, whoever merges this, **please squash the commits**.

**Comment by pablocarle** [comment]
@ifakhrutdinov I don't have merge permissions in this repository, pls squash & merge

**Review by JoeNemo — APPROVED** [review]
Approved.

**Pull request merged**
